src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java

```diff
@@ -123,55 +131,114 @@ protected void setState(State state)
+    {
+        ImmutableMap<TxnId, Command> ctxCommands = commandCache.getActive(preLoadContext.txnIds());
+        ImmutableMap<RoutableKey, CommandsForKey> ctxCommandsForKey = cfkCache.getActive(toRoutableKeys(preLoadContext.keys()));
+        return new PreExecuteContext()
+        {
+            @Override
+            public ImmutableMap<TxnId, Command> commands()
+            {
+                return ctxCommands;
+            }
+
+            @Override
+            public ImmutableMap<RoutableKey, CommandsForKey> commandsForKey()
+            {
+                return ctxCommandsForKey;
+            }
+
+            @Override
+            public Iterable<TxnId> txnIds()
+            {
+                return preLoadContext.txnIds();
+            }
+
+            @Override
+            public Seekables<?, ?> keys()
+            {
+                return preLoadContext.keys();
+            }
+        };
+    }
+
+    private static <K, V extends ImmutableState> void releaseResources(AccordStateCache.Instance<K, V> cache,
+                                                                       java.util.Map<K, ContextValue<V>> values)
+    {
+         values.forEach((key, value) -> {
+             cache.release(key, value.original(), value.current());
+         });
+    }
+
+    private void releaseResources()
+    {
+        releaseResources(commandStore.commandCache(), postExecuteContext.commands);
+        releaseResources(commandStore.commandsForKeyCache(), postExecuteContext.commandsForKey);
+    }
+
     protected void runInternal()
     {
-        SafeAccordCommandStore safeStore = commandStore.safeStore(context);
         switch (state)
         {
             case INITIALIZED:
                 state = State.LOADING;
             case LOADING:
-                if (!loader.load(context, this))
+                if (!loader.load(this::callback))
                     return;
 
                 state = State.RUNNING;
+                preExecuteContext = preExecuteContext(preLoadContext, commandStore.commandCache(), commandStore.commandsForKeyCache());
+                SafeCommandStore safeStore = commandStore.beginOperation(preExecuteContext);
                 result = apply(safeStore);
+                postExecuteContext = commandStore.completeOperation(safeStore);
 
                 state = State.SAVING;
             case SAVING:
             case AWAITING_SAVE:
-                boolean updatesPersisted = writer.save(context, this);
+                boolean updatesPersisted = writer.save(postExecuteContext, this::callback);
 
-                if (state != State.AWAITING_SAVE)
+                if (state == State.SAVING)
                 {
                     // with any updates on the way to disk, release resources so operations waiting
                     // to use these objects don't have issues with fields marked as unsaved
-                    context.releaseResources(commandStore);
+                    releaseResources();
```

Comment:
I think you need to move this before `writer.save` as this does checks for correctness... for example if 2 different operations ran concurrently (not in parallel) if we ever had 1 operation not see the mutation of the other one (bug) we fail populating the cache, but we currently apply that mutation to disk; feel we should validate these assumptions before making them durable?

Also this name is misleading, it's not "releasing" but "updating" the cache.  We do lower the ref count on the node, but we are also updating it... so maybe "updateAndDereferenceCache"?

Reply:
It no longer updates the cache (and when exactly that happens has been fluid), and the safetly checks performed now are mostly about the cache. Regarding naming I think releaseResources is clearer than dereferenceCache?